src/base/rss/utorrentrssdata.cpp

```diff
@@ -0,0 +1,218 @@
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "utorrentrssdata.h"
+
+#include <libtorrent/lazy_entry.hpp>
+
+#include <QDebug>
+#include <QFile>
+#include <QStringList>
+
+#ifdef QBT_USES_QT5
+#include <QRegularExpression>
+#else
+#include <QRegExp>
+#endif
+
+#include "rssfeed.h"
+#include "rssdownloadrule.h"
+#include "rssdownloadrulelist.h"
+#include "base/utils/string.h"
+
+using namespace Rss;
+using namespace libtorrent;
+using namespace Utils::String;
+
+namespace utorrent
+{
+    int POSTPONE_OPTS[] = { 0, 0, 0, 1, 2, 3, 4, 7, 14, 21, 30 };
+
+    enum EnabledFlags
+    {
+        ENABLED      = 1,
+        RAW_NAME     = 2,
+        MAX_PRIORITY = 4,
+        SMART_FILTER = 8,
+        PAUSED       = 16
+    };
+
+    enum Quality
+    {
+        QUALITY_ALL   = -1,
+        QUALITY_480p  = 0x10000,
+        QUALITY_720p  = 0x00800,
+        QUALITY_1080i = 0x01000,
+        QUALITY_1080p = 0x02000
+    };
+
+    void applyFilter(const Rss::DownloadRulePtr &rule, const libtorrent::lazy_entry &filter)
+    {
+        using namespace Utils::String;
+
+#ifdef QBT_USES_QT5
+        // Leading/Trailing *'s and *'s before/after '|'
+        const QRegularExpression STRIP_PATTERN("(?<=^|\\|)\\*+|\\*+(?=$|\\|)");
+#else
+        // Leading/Trailing *'s and *'s before '|'
+        // No lookbehind, so as close as it'll get without complicating things
+        const QRegExp STRIP_PATTERN("^\\*+|\\*+(?=$|\\|)");
+#endif
+
+        rule->setUseRegex(true);
+        rule->setMustNotContain(fromStdString(filter.dict_find_string_value("not_filter")).remove(STRIP_PATTERN).replace('*', ".*").replace('?', "."));
+
+        QString mustContain = fromStdString(filter.dict_find_string_value("filter")).remove(STRIP_PATTERN).replace('*', ".*").replace('?', ".");
+
+        QStringList qualityList;
+        int quality = filter.dict_find_int_value("quality");
+        if (quality != QUALITY_ALL) {
+            if (quality & QUALITY_480p)  qualityList << "480p";
+            if (quality & QUALITY_720p)  qualityList << "720p";
+            if (quality & QUALITY_1080i) qualityList << "1080i";
+            if (quality & QUALITY_1080p) qualityList << "1080p";
+        }
+
+        if (qualityList.empty())
+            rule->setMustContain(mustContain);
+        else
+            rule->setMustContain("(" % mustContain % ").*(" % qualityList.join("|") % ")");
+    }
+}
+
+UTorrentRssData::UTorrentRssData()
+    : m_rules(new DownloadRuleList)
+{
+}
+
+UTorrentRssData::~UTorrentRssData()
+{
+    if (m_rules) delete m_rules;
```

Comment:
Turns out I don't. I had thought I needed it in my original implementation of this class, and forgot to remove it after refactoring. It's been a while since I've done C++, I forgot it was safe to delete null pointers.